karakeep: module tweaks and added module test

tetov opened the pull request:
I started this PR with the intent to expose host and port as their own options but after going through similar modules I found that that is not typically done. I had already written a test and done some refactoring so I decided to open a PR anyways.

- Added a nixos test which mainly targets environment variables HOST and PORT.
  No non graphical way to generate api token found so not sure how to test more.

- Changed how the systemd environment is constructed using attribute merging

- environmentFiles added to replace environmentFile, which is hidden but still
  supported (with a deprecation warning). This fits well with how the env file
  generated in the init service unit is added to web and workers.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [Nixpkgs 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/doc/release-notes/rl-2511.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/doc/manual/release-notes/rl-2505.section.md) Nixpkgs Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
- [NixOS 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2511.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) NixOS Release notes)
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

tetov committed:
nixos/karakeep: refactor and test


- Added a nixos test which mainly targets environment variables HOST and PORT.
  No non graphical way to generate api token found so not sure how to test more.

- Changed how the systemd environment is constructed using attribute merging

- environmentFiles added to replace environmentFile, which is hidden but still
  supported (with a deprecation warning). This fits well with how the env file
  generated in the init service unit is added to web and workers.

tetov commented:
Requested review from module and package maintainer @three.

tetov commented:
@ofborg test karakeep

three reviewed (approved):
I appreciate the cleanup! I did have a question about how setting `services.karakeep.environmentFiles` from within the module works. Otherwise this looks good.

tetov committed:
nixos/karakeep: remove options that are better as freeform

tetov committed:
nixos/karakeep: use freeformType for extraEnvironment, refactor


DATA_DIR is protected as it is set to read-only:

error: The option `nodes.karakeep_server.services.karakeep.extraEnvironment.DATA_DIR' is read-only, but it's set multiple times. Definition values:
- In `/nix/store/AAA-source/nixos/modules/services/web-apps/karakeep.nix': "/var/lib/karakeep"
- In `/nix/store/AAA-source/nixos/tests/web-apps/karakeep.nix': "/something/else"

environmentFile is now deprecated properly, it gives a warning but still works (by turning the value into a list and assigning to environmentFiles):

evaluation warning: The option `services.karakeep.environmentFile' defined in `/nix/store/AAA-source/nixos/tests/web-apps/karakeep.nix' has been changed to `services.karakeep.environmentFiles' that has a different type. Please read `services.karakeep.environmentFiles' documentation and update your configuration accordingly.

tetov commented:
> I did have a question about how setting `services.karakeep.environmentFiles` from within the module works.

I found this behavior in another module, I'll try to find it. I thought it made sense to use the default behavior of option merging :).

HritwikSinghal commented:
Hi @tetov , [karakeep 0.25.0](https://github.com/karakeep-app/karakeep/releases/tag/v0.25.0) just got released. Can you also update the pkg to latest version if you know how its nix packaging works?

tetov commented:
@HritwikSinghal: https://github.com/NixOS/nixpkgs/pull/416487 :)

tetov commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review-gha`](https://github.com/Defelo/nixpkgs-review-gha)

Command: `nixpkgs-review pr 412414`

Logs: https://github.com/tetov/nixpkgs-review-gha/actions/runs/15757816518


---
### `x86_64-linux` (sandbox = true)
<details>
  <summary>:fast_forward: 2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>


---
### `aarch64-linux` (sandbox = true)
<details>
  <summary>:fast_forward: 2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>

mdaniels5757 commented:
@ofborg build karakeep karakeep.passthru.tests

tetov commented:
Fixed merge conflict.
